Spigot-Server-Patches/0329-PlayerShootProjectileEvent.patch

```diff
@@ -0,0 +1,228 @@
++            }
++            // Paper end
+         }
+ 
+         entityhuman.b(StatisticList.b((Item) this));
+diff --git a/src/main/java/net/minecraft/server/ItemLingeringPotion.java b/src/main/java/net/minecraft/server/ItemLingeringPotion.java
+index aadf9e88d..1863d1105 100644
+--- a/src/main/java/net/minecraft/server/ItemLingeringPotion.java
++++ b/src/main/java/net/minecraft/server/ItemLingeringPotion.java
+@@ -10,14 +10,27 @@ public class ItemLingeringPotion extends ItemPotion {
+ 
+     public InteractionResultWrapper<ItemStack> a(World world, EntityHuman entityhuman, EnumHand enumhand) {
+         ItemStack itemstack = entityhuman.b(enumhand);
+-        ItemStack itemstack1 = entityhuman.abilities.canInstantlyBuild ? itemstack.cloneItemStack() : itemstack.cloneAndSubtract(1);
+-
+-        world.a((EntityHuman) null, entityhuman.locX, entityhuman.locY, entityhuman.locZ, SoundEffects.dL, SoundCategory.NEUTRAL, 0.5F, 0.4F / (ItemLingeringPotion.j.nextFloat() * 0.4F + 0.8F));
++        // Paper start - moved down
++        //ItemStack itemstack1 = entityhuman.abilities.canInstantlyBuild ? itemstack.cloneItemStack() : itemstack.cloneAndSubtract(1);
++        //
++        //world.a((EntityHuman) null, entityhuman.locX, entityhuman.locY, entityhuman.locZ, SoundEffects.dL, SoundCategory.NEUTRAL, 0.5F, 0.4F / (ItemLingeringPotion.j.nextFloat() * 0.4F + 0.8F));
++        // Paper end
+         if (!world.isClientSide) {
+-            EntityPotion entitypotion = new EntityPotion(world, entityhuman, itemstack1);
++            EntityPotion entitypotion = new EntityPotion(world, entityhuman, itemstack.cloneItemStack()); // Paper - clone itemstack
+ 
+             entitypotion.a(entityhuman, entityhuman.pitch, entityhuman.yaw, -20.0F, 0.5F, 1.0F);
+-            world.addEntity(entitypotion);
++            // Paper start
++            com.destroystokyo.paper.event.entity.PlayerShootProjectileEvent event = new com.destroystokyo.paper.event.entity.PlayerShootProjectileEvent((org.bukkit.entity.Player) entityhuman.getBukkitEntity(), org.bukkit.craftbukkit.inventory.CraftItemStack.asCraftMirror(itemstack), (org.bukkit.entity.Projectile) entitypotion.getBukkitEntity());
++            if (event.callEvent() && world.addEntity(entitypotion)) {
++                if (event.getConsumeStack() && !entityhuman.abilities.canInstantlyBuild) {
++                    entitypotion.setItem(itemstack.cloneAndSubtract(1));
++                } else ((EntityPlayer) entityhuman).getBukkitEntity().updateInventory();
++
++                world.a((EntityHuman) null, entityhuman.locX, entityhuman.locY, entityhuman.locZ, SoundEffects.dL, SoundCategory.NEUTRAL, 0.5F, 0.4F / (ItemLingeringPotion.j.nextFloat() * 0.4F + 0.8F));
++            } else if (entityhuman instanceof EntityPlayer) {
++                ((EntityPlayer) entityhuman).getBukkitEntity().updateInventory();
++            }
++            // Paper end
+         }
+ 
+         entityhuman.b(StatisticList.b((Item) this));
```

Comment:
I'm not sure we should fire statistics for cancelled events? same for the rest of the changed items.